src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs

```diff
@@ -27,7 +28,7 @@ namespace NuGet.SolutionRestoreManager
     [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
     [ProvideAutoLoad(VSConstants.UICONTEXT.SolutionExists_string)]
     [Guid(PackageGuidString)]
-    public sealed class RestoreManagerPackage : AsyncPackage
+    public sealed class RestoreManagerPackage : AsyncPackage, IVsUpdateSolutionEvents4
```

Comment:
I don't think that means it has to be here. In fact since other folks get pointers to your Package, this means they can cast that pointer to that interface and invoke you. That's probably not going to become a problem, but information hiding usually provides benefits where you don't expect them.
Just have a private nested class within this Package class (or another independent internal class) that implements this interface, new it up in your Package constructor and pass that instance to the Advise method.